### PR TITLE
spring boot security, roles, PreAuthorize access

### PR DESCRIPTION
**Pull request by OlgaStarikova**

### CODE REVIEW

File: src/main/resources/db/changelog/changes/04-create-user-role-table.yml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+                    nullable: false
+                    unique: true
+                    uniqueConstraintName: user_id-role_id_unique
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
+                    nullable: false
+                    unique: true
+                    uniqueConstraintName: user_id-role_id_unique
```
Comment (kerrrusha) [inline src/main/resources/db/changelog/changes/04-create-user-role-table.yml:22]: what about foreign keys?

**Review by kerrrusha — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/onlinebookstore/config/SecurityConfig.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.security.web.SecurityFilterChain;
+
+@RequiredArgsConstructor
+@Configuration
+@EnableWebSecurity
+@EnableMethodSecurity
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/config/SecurityConfig.java:20]: Do you need both annotation here `@EnableWebSecurity and @EnableMethodSecurity`

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 3]
```diff
@@ -31,43 +32,50 @@ public class BookController {
...
         bookService.deleteBook(id);
     }
 
     @PutMapping("/{id}")
-    @Operation(summary = "Update the book", description = "Update the book by Id"
-            + "Params: id = Id of the book")
+    @Operation(summary = "Update the book", description = "Update the book by Id."
+            + "Params: id = Id of the book. Available for admins.")
+    @PreAuthorize("hasAuthority('ADMIN')")
     public BookDto updateBook(@PathVariable Long id, @RequestBody CreateBookRequestDto requestDto) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java:71]:
```suggestion
    public BookDto updateBook(@PathVariable Long id, @RequestBody @Valid CreateBookRequestDto requestDto) {
```

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 4]
```diff
@@ -31,43 +32,50 @@ public class BookController {
...
         return bookService.updateBook(id, requestDto);
     }
 
     @GetMapping("/search")
     @Operation(summary = "Search books", description = "Get a list of books by parameters"
-            + "Params: String[] titles, String[] authors")
+            + "Params: String[] titles, String[] authors. Available for registered users.")
+    @PreAuthorize("hasAuthority('USER')")
     public List<BookDto> search(BookSearchParameters searchParameters) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java:79]:
```suggestion
    public List<BookDto> search(BookSearchParameters searchParameters, @ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 5]
```diff
@@ -31,43 +32,50 @@ public class BookController {
     @GetMapping
     @Operation(summary = "Get a list of books", description = "Get a list of all available books."
             + "Params(optional): page = page number, size = count of books in one page,"
-            + " namefield = field for sorting")
+            + " namefield = field for sorting. Available for registered users.")
+    @PreAuthorize("hasAuthority('USER')")
     public List<BookDto> getAll(Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java:37]:
```suggestion
    public List<BookDto> getAll(@ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/com/example/onlinebookstore/model/Role.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.springframework.security.core.GrantedAuthority;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/model/Role.java:13]:
```suggestion
@Entity
@Data
```
or add necessary annotations

File: src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java [hunk 7]
```diff
@@ -22,6 +25,12 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
                     + requestDto.email()
                     + " is already registered");
         }
-        return userMapper.toDto(userRepository.save(userMapper.toModel(requestDto)));
+        User user = new User();
+        user.setPassword(passwordEncoder.encode(requestDto.password()));
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java:29]: Do we need to set Role.USER here?

File: src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java [hunk 8]
```diff
@@ -22,6 +25,12 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto) {
                     + requestDto.email()
                     + " is already registered");
         }
-        return userMapper.toDto(userRepository.save(userMapper.toModel(requestDto)));
+        User user = new User();
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/onlinebookstore/service/impl/UserServiceImpl.java:28]:
```suggestion
        User user = userMapper.toUser(requestDto);
        user.setPassword(passwordEncoder.encode(requestDto.password()));
        user.setRoles(...
        return userMapper.toDto(userRepository.save(user));
```

File: src/main/resources/db/changelog/changes/04-create-user-role-table.yml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+      id: create-user-role-table
+      author: Olya Starykova
+      changes:
+        - createTable:
+            tableName: users_roles
+            columns:
+              - column:
+                  name: user_id
+                  type: bigint
+                  constraints:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-user-role-table.yml:12]:
```suggestion
                  constraints:
                    foreignKeyName: fk_users_roles_users
                    referencedTableName: users
                    referencedColumnNames:  id
                    nullable:  false
```

File: src/main/resources/db/changelog/changes/04-create-user-role-table.yml [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+                    nullable: false
+                    unique: true
+                    uniqueConstraintName: user_id-role_id_unique
+              - column:
+                  name: role_id
+                  type: bigint
+                  constraints:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-create-user-role-table.yml:19]:
```suggestion
                  constraints:
                    foreignKeyName: fk_users_roles_roles
                    referencedTableName: roles
                    referencedColumnNames:  id
                    nullable:  false
```

File: src/main/resources/db/changelog/changes/08-addFK-user-role-table.yml [hunk 11]
```diff
@@ -0,0 +1,23 @@
+databaseChangeLog:
+  -  changeSet:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/08-addFK-user-role-table.yml:2]: You actually don't need this changelog

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's improve your solution!

**Review by Elena-Bruyako — APPROVED** [review]
Good job!

**Pull request merged**